api/v1beta1/azuremachine_validation.go

```diff
@@ -378,3 +392,37 @@ func ValidateDiagnostics(diagnostics *Diagnostics, fieldPath *field.Path) field.
 
 	return allErrs
 }
+
+// ValidateConfidentialCompute validates the configuration options when the machine is a Confidential VM.
+func ValidateConfidentialCompute(managedDisk *ManagedDiskParameters, profile *SecurityProfile, fieldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	var securityEncryptionType SecurityEncryptionType
+
+	if managedDisk != nil && managedDisk.SecurityProfile != nil {
+		securityEncryptionType = managedDisk.SecurityProfile.SecurityEncryptionType
+	}
+
+	if profile != nil && securityEncryptionType != "" {
+		if profile.EncryptionAtHost != nil && *profile.EncryptionAtHost && securityEncryptionType == SecurityEncryptionTypeDiskWithVMGuestState {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("EncryptionAtHost"), profile.EncryptionAtHost,
+				"EncryptionAtHost cannot be set to 'true' when securityEncryptionType is set to DiskWithVMGuestState"))
+		}
+
+		if profile.VirtualizedTrustedPlatformModule == "" || profile.VirtualizedTrustedPlatformModule == VirtualizedTrustedPlatformModulePolicyDisabled {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("VirtualizedTrustedPlatformModule"), profile.VirtualizedTrustedPlatformModule,
+				fmt.Sprintf("VirtualizedTrustedPlatformModule should be set to '%s' when securityEncryptionType is defined",
```

Comment:
I believe this is not accurate, Azure allows you to deploy with vTPM disabled (kinda pointless though

Reply:
ah, I might be mixing up CVM and TL. have you tested this out with Azure VMs?

Reply:
In that case, I incorrectly blindly followed the Azure Terraform Provider [docs](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_virtual_machine#security_encryption_type), but I will definitely try out what you suggested and revert this change.

Reply:
Trying to create a `ConfidentialVM` with `SecurityTypeEncryption: VMGuestStateOnly | DiskWithVMGuestState` and `vTpmEnabled: false` fails with the following error:

```shell
> az vm create \
  --resource-group aResourceGroup \
  --name myConfidentialVM \
  --generate-ssh-keys \
  --size Standard_DC4as_v5 \
  --admin-username aUserame \
  --admin-password aPassword \
  --enable-vtpm false \
  --image "Canonical:0001-com-ubuntu-confidential-vm-focal:20_04-lts-cvm:latest" \
  --public-ip-sku Standard \
  --security-type ConfidentialVM \
  --os-disk-security-encryption-type VMGuestStateOnly \
  --enable-secure-boot false

ERROR: {"status":"Failed","error":{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.","details":[{"code":"BadRequest","message":"{\r\n  \"error\": {\r\n    \"code\": \"BadRequest\",\r\n    \"message\": \"Confidential VM or Virtual Machine Scale Set VM with 'managedDisk.securityProfile.securityEncryptionType' set as VMGuestStateOnly should have securityProfile.uefiSettings.vtpmEnabled set to true.\"\r\n  }\r\n}"}]}}
```

```shell
> az vm create \
  --resource-group aResourceGroup \
  --name myConfidentialVM \
  --generate-ssh-keys \
  --size Standard_DC4as_v5 \
  --admin-username aUsername \
  --admin-password aPassword \
  --enable-vtpm false \
  --image "Canonical:0001-com-ubuntu-confidential-vm-focal:20_04-lts-cvm:latest" \
  --public-ip-sku Standard \
  --security-type ConfidentialVM \
  --os-disk-security-encryption-type DiskWithVMGuestState \
  --enable-secure-boot false \

ERROR: {"status":"Failed","error":{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.","details":[{"code":"BadRequest","message":"{\r\n  \"error\": {\r\n    \"code\": \"BadRequest\",\r\n    \"message\": \"Confidential VM or Virtual Machine Scale Set VM with 'managedDisk.securityProfile.securityEncryptionType' set as DiskWithVMGuestState should have both values of securityProfile.uefiSettings.secureBootEnabled and securityProfile.uefiSettings.vtpmEnabled set to true.\"\r\n  }\r\n}"}]}}
```

Reply:
yeah, I think you got this right for CVM. apologies for confusion :)

for clarity -- vTPM and SecureBoot I believe are both required for CVM and both optional for trusted launch.

it looks like this func is called only for securityEncryptionType != "", but is that exclusive to CVM? I haven't tried securityEncryptionType with TL myself I think?

either way I think at worst we "under validate" vs disallow valid settings, so we can always tweak later.